src/Compilers/CSharp/Portable/BoundTree/UnboundLambda.cs

```diff
@@ -972,15 +972,41 @@ private static int CanonicallyCompareDiagnostics(Diagnostic x, Diagnostic y)
 
             int codeCompare = xCode.CompareTo(yCode);
 
-            // ToString fails for a diagnostic with an error code that does not prevent successful delegate conversion.
-            // Also, the order doesn't matter, since all such diagnostics will be dropped.
-            if (!ErrorFacts.PreventsSuccessfulDelegateConversion(xCode) || !ErrorFacts.PreventsSuccessfulDelegateConversion(yCode))
+            // Optimization: don't bother 
+            if (codeCompare != 0)
+                return codeCompare;
+
+            // According to a previous comment:
+            //
+            // "ToString fails for a diagnostic with an error code that does not prevent successful delegate conversion.
+            //  Also,the order doesn't matter, since all such diagnostics will be dropped."
+            //
+            // Currently we no longer call Diagnostic.ToString(), but call ToString() on the arguments.
+            // However since we were unable to reproduce this exception, it is possible it came from the arguments.
+            // Besides, the arguments can sometimes be Symbols, for which ToString is an expensive operation, so this is probably an optimization.
+            if (!ErrorFacts.PreventsSuccessfulDelegateConversion(xCode))
             {
                 return codeCompare;
             }
 
-            // Optimization: don't bother 
-            return codeCompare == 0 ? string.CompareOrdinal(x.ToString(), y.ToString()) : codeCompare;
+            if (x.Arguments is null && y.Arguments is null)
+                return 0;
+            if (x.Arguments is null)
+                return -1;
+            if (y.Arguments is null)
+                return 1;
```

Comment:
Perhaps treat `null` and empty as equivalent.
```
int nx = x.Arguments?.Count ?? 0;
int ny = y.Arguments?.Count ?? 0;
for (int i = 0, n = Min(nx, ny); ...)
{
    ...
}
return nx - ny;
```

Reply:
Once we move to NRT, that will probably not satisfy the flow analysis, so it might be best to leave it as is.